umap/static/umap/js/modules/autocomplete.js

```diff
@@ -0,0 +1,317 @@
+        }
+        break
+    }
+  }
+
+  onKeyUp(e) {
+    const special = [
+      'Tab',
+      'Enter',
+      'ArrowLeft',
+      'ArrowRight',
+      'ArrowDown',
+      'ArrowUp',
+      'Meta',
+      'Shift',
+      'Alt',
+      'Control',
+    ]
+    if (!special.includes(e.key)) {
+      this.search()
+    }
+  }
+
+  onBlur() {
+    setTimeout(() => this.hide(), 100)
+  }
+
+  clear() {
+    this.results = []
+    this.current = null
+    this.cache = ''
+    this.container.innerHTML = ''
+  }
+
+  hide() {
+    this.clear()
+    this.container.style.display = 'none'
+    this.input.value = ''
+  }
+
+  setChoice(choice) {
+    console.log(choice)
+    choice = choice || this.results[this.current]
+    if (choice) {
+      this.input.value = choice.item.label
+      this.options.on_select(choice)
+      this.displaySelected(choice)
+      this.hide()
+      if (this.options.callback) {
+        this.options.callback.bind(this)(choice)
+      }
+    }
+  }
+
+  createResult(item) {
+    const el = DomUtil.element({
+      tagName: 'li',
+      parent: this.container,
+      textContent: item.label,
+    })
+    const result = {
+      item: item,
+      el: el,
+    }
+    DomEvent.on(el, 'mouseover', () => {
+      this.current = result
+      this.highlight()
+    })
+    DomEvent.on(el, 'mousedown', () => this.setChoice())
+    return result
+  }
+
+  resultToIndex(result) {
+    let out = null
+    this.results.forEach((item, index) => {
+      if (item.item.value == result.item.value) {
+        out = index
+        return
+      }
+    })
+    return out
+  }
+
+  handleResults(data) {
+    this.clear()
+    this.container.style.display = 'block'
+    this.resizeContainer()
+    data.forEach((item) => {
+      this.results.push(this.createResult(item))
+    })
+    this.current = 0
+    this.highlight()
+    //TODO manage no results
+  }
+
+  highlight() {
+    this.results.forEach((result, index) => {
+      if (index === this.current) DomUtil.addClass(result.el, 'on')
+      else DomUtil.removeClass(result.el, 'on')
+    })
+  }
+
+  getLeft(el) {
+    let tmp = el.offsetLeft
+    el = el.offsetParent
+    while (el) {
+      tmp += el.offsetLeft
+      el = el.offsetParent
+    }
+    return tmp
+  }
+
+  getTop(el) {
+    let tmp = el.offsetTop
+    el = el.offsetParent
+    while (el) {
+      tmp += el.offsetTop
+      el = el.offsetParent
+    }
+    return tmp
+  }
+}
+
+class BaseAjax extends BaseAutocomplete {
+  constructor(el, options) {
+    super(el, options)
+    // FIXME when UI elements are moved to modules
+    const ui = new U.UI(document.querySelector('header'))
+    this.server = new ServerRequest(ui)
+  }
+  optionToResult(option) {
+    return {
+      value: option.value,
+      label: option.innerHTML,
+    }
+  }
+
+  async search() {
+    let val = this.input.value
+    if (val.length < this.options.minChar) {
+      this.clear()
+      return
+    }
+    if (`${val}` === `${this.cache}`) return
```

Comment:
Weird comparison, why don't we compare directly the variables?

Reply:
No idea, I removed. I'm not totally why this check if for, actually. So let's fix and see :p